services/deployment-agent/src/simcore_service_deployment_agent/git_url_watcher.py

```diff
@@ -16,53 +17,74 @@
 
 
-async def _git_checkout_file(directory: Path, file_path: Path):
-    await _git_checkout_files(directory, [file_path])
+async def _git_get_current_sha(directory: Path) -> str:
+    cmd = f"cd {directory} && git rev-parse --short HEAD"
+    sha = await run_cmd_line(cmd)
+    return sha.strip("\n")
 
+async def _git_clean_repo(directory: Path):
+    cmd = f"cd {directory} && git clean -dxf"
+    await run_cmd_line(cmd)
 
-async def _git_checkout_repo(directory: Path):
-    cmd = "cd {directory} && git checkout HEAD".format(directory=directory)
+async def _git_checkout_files(directory: Path, paths: List[Path], tag: str = None):
+    if not tag:
+        tag = "HEAD"
+    cmd = f"cd {directory} && git checkout {tag} {' '.join(paths)}"
     await run_cmd_line(cmd)
 
+async def _git_checkout_repo(directory: Path, tag: str = None):
+    await _git_checkout_files(directory, [], tag)
 
 async def _git_pull_files(directory: Path, paths: List[Path]):
-    cmd = "cd {directory} && git checkout FETCH_HEAD {path}".format(directory=directory,
-                                                                    path=" ".join(paths))
+    cmd = f"cd {directory} && git checkout FETCH_HEAD {' '.join(paths)}"
     await run_cmd_line(cmd)
 
 
 async def _git_pull(directory: Path):
-    cmd = "cd {directory} && git pull".format(directory=directory)
+    cmd = f"cd {directory} && git pull"
     await run_cmd_line(cmd)
 
-
 async def _git_fetch(directory: Path):
-    cmd = "cd {directory} && git fetch --prune".format(directory=directory)
+    cmd = f"cd {directory} && git fetch --prune --tags"
     await run_cmd_line(cmd)
 
+async def _git_get_latest_matching_tag(directory: Path, regexp: str) -> List[str]:
+    cmd = f"cd {directory} && git --no-pager describe --tags --always `git rev-list --tags`"
+    all_tags = await run_cmd_line(cmd)
+    if not all_tags:
+        return []
+    list_tags = all_tags.split("\n")
+    reg_object = re.compile(regexp)
+    for tag in list_tags:
+        if reg_object.fullmatch(tag):
+            return tag
+    return ""
+
+async def _git_get_current_matching_tag(directory: Path, regexp: str) -> str:
+    cmd = f"cd {directory} && git --no-pager describe --tags --always"
+    current_tag = (await run_cmd_line(cmd)).strip("\n")
+    match = re.fullmatch(regexp, current_tag)
+    if match:
+        return current_tag
+    return ""
 
 async def _git_diff_filenames(directory: Path) -> str:
-    cmd = "cd {directory} && git --no-pager diff --name-only FETCH_HEAD".format(
-        directory=directory)
+    cmd = f"cd {directory} && git --no-pager diff --name-only FETCH_HEAD"
     modified_files = await run_cmd_line(cmd)
     return modified_files
 
-async def _git_get_logs(directory: Path, branch: str) -> str:
-    cmd = "cd {directory} && git log --oneline {branch}..origin/{branch}".format(
-        directory=directory, branch=branch)
+async def _git_get_logs(directory: Path, branch1: str, branch2: str) -> str:

```

Comment:
I wonder if you could copy this collection of ``_git_..`` helper functions and add them e.g. in ``servicelib``

Reply:
well maybe.. don't know if we need this elsewhere at the moment.